_episodes/l3_03_test_fixture.md

```diff
@@ -0,0 +1,505 @@
+                     0.2*30000 + \
+                     0.5*(salary_with_bonus-50000);
+    }
+}
+
+// Function to calculate the ney salary.
+void Employee::calcNetSalary() 
+{
+    net_salary = base_salary + net_bonus - tax_amount;
+}
+```
+With this code, we now have the necessary fragments to test our `Employee` class. Let us see this in action.
+
+## 3. Unit tests for our employee class without test fixture
+
+In order to clearly demonstrate why a test fixture would be needed, we first write some tests for our employee class without using a fixture. This will help us to understand why does a fixture is useful and how to use it.
+
+For this subsection, let us assume that we are checking two functionalities of our employee class (for code, please see [1_employeetest.cpp](../code/Chapter3/1_employeetest.cpp)) which are:-
+
+1. We can set the name of employee correctly.
+2. We can set the age correctly.
+
+The code for these two tests (which is entirely based on what we learn in second module of this course) is given below.
+
+```cpp
+// Test if we can set the name of an employee.
+TEST(EmployeeTest, CanSetName) {
+    Employee employee{"John", 25, 10000, 5, 1000};
+    employee.setName("John Doe");
+    EXPECT_EQ(employee.getName(), "John Doe");
+}
+
+// Test if we can set the age of an employee.
+TEST(EmployeeTest, CanSetAge) {
+    Employee employee{"John", 25, 10000, 5, 1000};
+    employee.setAge(30);
+    EXPECT_EQ(employee.getAge(), 30);
+}
+```
+
+While the above test solve our problem, there is a problem of code duplication and object creation for each test. As we can see in each test that we have to create an instance of employee class by using the statement `Employee employee{"John", 25, 10000, 5, 1000};`. This is against the `DRY (Don't Repeat Yourself)' https://en.wikipedia.org/wiki/Don%27t_repeat_yourself.
+
+Moreover, all our tests depend on the same `employee` class. Therefore, it makes sense to create an instance of `employee` at one place and let the GoogleTest manage the creation of the instance for each test case. Let us see this in action in next section.
+
+## 4. Test fixture for our employee class
+
+Now that we know why do we need a test fixture, let us fist learn about the basic syntax of the test fixture in GoogleTest and then write the code for it.
+
+In GooglTest, a test fixture is created by writing another class which is derived from `::testing::Test` using `public` [access specifier](https://en.cppreference.com/w/cpp/language/access). The general syntax is as shown below
+
+```cpp
+class Your_test_fixture_class_name : public::testing::Test {
+    public:
+        ClassUnderTest publicInstance;
+
+    protected:
+        ClassUnderTest protectedInstance;
+
+    private:
+        ClassUnderTest privateInstance;
+};
+```
+
+Please note that you do not need to use all three access specifiers `public`, `private` and `protected`  defined above. The choice would depend on the following:-
```

Comment:
Do we really need to explain `public`, `protected` and `private` here? One might hope the reader knows about access specifiers. I'm not sure the explanations of the latter two are important for the task. Maybe briefly mention it if it's critical.

Reply:
These access specifiers have been defined in the context of test fixtures (although the meaning remains same). When I started learning, many of the texts or online resources were using `public` for declaring the class under test. I was under the impression that this is the only choice that is available (Sorry, my lack of understanding).

Then I felt like that some students may also feel the same. This is why I introduced all 3 here and specifically  mentioned  that all are OK and allowed and why to choose one over another.

I do not think it will take too much of time during the course and I think it is useful.

If you still want me to delete this section, I am happy to do that.

Reply:
I'd expect participants of this course to understand access specifiers, but I agree that these are defined in the context of test fixtures, so should be good to keep.

Reply:
I'm with Dan. While specifiers are probably well known, here they are explained in the context of fixtures, which add a small twist to them.